engine/src/main/java/org/terasology/rendering/world/dag/ShadowMapNode.java

```diff
@@ -0,0 +1,171 @@
+    // FIXME: remove reference to WorldRendererImpl
+    private WorldRendererImpl worldRenderer;
+    private BackdropProvider backdropProvider;
+
+
+    // FIXME: unnecessary arguments must be eliminated
+    public ShadowMapNode(Material shader,
+                         Camera playerCamera,
+                         WorldRendererImpl worldRenderer,
+                         BackdropProvider backdropProvider,
+                         RenderingConfig renderingConfig,
+                         GraphicState state) {
+
+        this.shader = shader;
+        this.playerCamera = playerCamera;
+        this.worldRenderer = worldRenderer;
+        this.backdropProvider = backdropProvider;
+        this.renderingConfig = renderingConfig;
+        this.state = state;
+
+    }
+
+
+    @Override
+    public void initialise() {
+        // TODO: shader and fbo shall be initialised here
+        // TODO: using ShaderParametersShadowMap here, instead of storing in ShaderManagerLwjgl
+    }
+
+    @Override
+    public void preRender(float update) {
+        // positionShadowMapCamera()
+
+        // Shadows are rendered around the player so...
+        Vector3f lightPosition = new Vector3f(playerCamera.getPosition().x, 0.0f, playerCamera.getPosition().z);
+
+        // Project the shadowMapCamera position to light space and make sure it is only moved in texel steps (avoids flickering when moving the shadowMapCamera)
+        float texelSize = 1.0f / renderingConfig.getShadowMapResolution();
+        texelSize *= 2.0f;
+
+        camera.getViewProjectionMatrix().transformPoint(lightPosition);
+        lightPosition.set(TeraMath.fastFloor(lightPosition.x / texelSize) * texelSize, 0.0f, TeraMath.fastFloor(lightPosition.z / texelSize) * texelSize);
+        camera.getInverseViewProjectionMatrix().transformPoint(lightPosition);
+
+        // ... we position our new shadowMapCamera at the position of the player and move it
+        // quite a bit into the direction of the sun (our main light).
+
+        // Make sure the sun does not move too often since it causes massive shadow flickering (from hell to the max)!
+        float stepSize = 50f;
+        Vector3f sunDirection = backdropProvider.getQuantizedSunDirection(stepSize);
+
+        Vector3f sunPosition = new Vector3f(sunDirection);
+        sunPosition.scale(256.0f + 64.0f);
+        lightPosition.add(sunPosition);
+
+        camera.getPosition().set(lightPosition);
+
+        // and adjust it to look from the sun direction into the direction of our player
+        Vector3f negSunDirection = new Vector3f(sunDirection);
+        negSunDirection.scale(-1.0f);
+
+        camera.getViewingDirection().set(negSunDirection);
+
+        camera.update(update);
+
+    }
+
+
+    @Override
+    public void render() {
+        // TODO: removing this assignment
+        fbo = state.buffers.sceneShadowMap; // added here since graphic settings can be modified and a new fbo can be generated.
```

Comment:
Ok, two things to do immediately:
1. Do not obtain the buffers from GraphicState. Get them from FrameBuffersManager, their actual home.
2. Do not hotlink the (originally private) buffer structure: use getFBO(String name)

Something for later: this fetches the FBO once per frame. Ideally:
1. the ShadowMapNode is responsible for generating the FBO and registering it with the FrameBuffersManager. This would require a "putFBO()" method in the manager - it's not there right now.
2. the ShadowMapNode should be responsible for monitoring the renderConfig and regenerate the FBOs on shadow map size changes. I think this should be handled by making the renderConfig observable, so that observer are notified of any changes or specific changes. 
3. the ShadowMapNode should also dispose of the FBO when the node itself gets disposed. 

More medium term, this makes me think that there is scope for a DAGNode.onRemoval() method and a separate DAGNode.onDeletion() method. This would allow for a node to be quickly re-inserted into the graph, it's structures and data not yet disposed. This way a user could quickly switch back&forth between having and not having the shadows, i.e. toggling back&forth the option in the config menu. Then, when the user leaves the menu, the node is either in the graph or gets deleted.